src/Hedgehog/Plutus/Tx.hs

```diff
@@ -0,0 +1,600 @@
+                _ -> False
+            )
+    )
+    (== pkh)
+
+-- | as balanceTx but with additional predicates about which TxOuts and PubKeyHashs can be used
+balanceTxWhere ::
+  TxContext ->
+  Tx 'Unbalanced ->
+  (Plutus.TxOut -> Bool) ->
+  (Plutus.PubKeyHash -> Bool) ->
+  Maybe (Hedgehog.Gen (Tx 'Balanced))
+balanceTxWhere context tx predTxout predPkh = do
+  let mock = mockchain context
+  (valueIn, valueOut) <- getValueInAndOut context tx
+  -- TODO we may want to change this to something like
+  -- Either FailReason (Tx 'Balanced)
+  -- to distinguish things like failure by tx lookup
+  -- and failure by not enough value at keys
+  let extra = posDif valueIn valueOut
+  let deficit = posDif valueOut valueIn
+  let alreadySpending txOut =
+        txOut
+          `elem` catMaybes
+            [ Map.lookup (txInRef txin) (Model.mockUtxos mock)
+            | txin <- Set.toList $ txInputs tx
+            ]
+  spendGen <- spendWhere mock deficit (liftM2 (&&) (not . alreadySpending) predTxout)
+  let payoutKeys = filter predPkh $ Map.keys $ Model.mockUsers mock
+  guard $ not $ null payoutKeys
+  pure $ do
+    Spend outRefs moreChange <- spendGen
+    payOutKey <- Gen.element payoutKeys
+    let change = extra <> moreChange
+    let checked =
+          confirmBalanced context $
+            tx
+              <> mempty
+                { txInputs =
+                    Set.fromList $
+                      Vector.toList $
+                        (`TxIn` Nothing) <$> outRefs
+                , txOutputs =
+                    pure $
+                      TxOut
+                        { txOutAddress = Plutus.pubKeyHashAddress payOutKey
+                        , txOutValue = change
+                        , txOutDatum = Nothing
+                        , txOutReferenceScript = Nothing
+                        }
+                }
+    case checked of
+      Just tx -> pure tx
+      Nothing -> error "balanceTx produced an unbalanced tx"
+
+getBalance :: TxContext -> Tx b -> Maybe Value
+getBalance context tx =
+  case (protocol, coreTx) of
+    (Model.AlonzoParams params, Alonzo coreTx) ->
+      cont @Alonzo.Era params coreTx
+    (Model.BabbageParams params, Babbage coreTx) ->
+      cont @Babbage.Era params coreTx
+    -- this error should be unreachable
+    _ -> error "era mismatch between balance and toLedgerTx"
+  where
+    cont ::
+      forall era.
+      ( Core.EraTx era
+      , Class.IsCardanoTx era
+      , CLI era
+      , ShelleyEraTxBody era
+      , Core.Value era ~ MV.MaryValue StandardCrypto
+      ) =>
+      Core.PParams era ->
+      Core.Tx era ->
+      Maybe Value
+    cont params coreTx = do
+      Right utxo <- pure $ Class.toUtxo mempty networkId []
+      -- TODO what should the utxo be for this?
+      pure $
+        maryToPlutus $
+          evaluateTransactionBalance @era
+            params
+            utxo
+            (const True) -- TODO is this right?
```

Comment:
I think the quick and dirty answer is that it doesn't matter right now, since staking doesn't work on psm. If/when that changes, this will become a problem, amongst many other things.